src/mlmc/hdf/level_group.py

```diff
@@ -0,0 +1,320 @@
+                chunks=True)
+            # Set data
+        else:
+            dataset = self._level_group['failed_ids']
+            dataset.resize((len(failed_samples),))
+            dataset[()] = list(failed_samples)
+
+    def _append_dataset(self, dataset_name, values):
+        """
+        Append values to existing dataset
+        :param dataset_name: Path to dataset
+        :param values: list of values (tuple or single value)
+        :return: None
+        """
+        if dataset_name not in self._level_group:
+            raise Exception("Dataset doesn't exist")
+        # Resize dataset
+        self._level_group[dataset_name].resize(self._level_group[dataset_name].shape[0] +
+                                               len(values), axis=0)
+        # Append new values to the end of dataset
+        self._level_group[dataset_name][-len(values):] = values
+
+    def scheduled(self):
+        """
+        Read level dataset with scheduled samples
+        :return: generator, each item is in form (Sample(), Sample())
+        """
+        # Generator of scheduled dataset values
+        scheduled_dset = self._level_group.get(LevelGroup.DATASET_NAMES['scheduled'], None)
+
+        if scheduled_dset is None:
+            return
+
+        # Create fine and coarse samples
+        for sample_id, (fine, coarse) in enumerate(scheduled_dset):
+            yield (Sample(sample_id=sample_id,
+                          directory=fine[0].decode('UTF-8'),
+                          job_id=fine[1].decode('UTF-8'),
+                          prepare_time=fine[2], queued_time=fine[3]),
+                   Sample(sample_id=sample_id,
+                          directory=coarse[0].decode('UTF-8'),
+                          job_id=coarse[1].decode('UTF-8'),
+                          prepare_time=coarse[2], queued_time=coarse[3]))
+
+    def collected(self):
+        """
+        Read all level datasets with collected data (reading dataset values via self._dataset_values() method),
+        create fine and coarse samples as Sample() instances
+        :return: generator; one item is tuple (Sample(), Sample())
+        """
+        # Collected ids dataset exists in Level group
+        if LevelGroup.DATASET_NAMES['collected_ids'] not in self._level_group:
+            return
+
+        # Number of collected samples
+        num_collected_samples = self._level_group[LevelGroup.DATASET_NAMES['collected_ids']].len()
+
+        # Collected data as dictionary according to lookup table COLLECTED_ATTRS
+        # dictionary format -> {COLLECTED_ATTRS key: corresponding dataset values generator, ...}
+        collected_data = {sample_attr_name: self._dataset_values('/'.join([self._level_group.name, dset_params['name']]))
+                          for sample_attr_name, dset_params in LevelGroup.COLLECTED_ATTRS.items()}
+
+        # For each item create two Sample() instances - fine sample and coarse sample
+        for _ in range(num_collected_samples):
```

Comment:
This seems to pass through the dataset one time for every attribute, bad idea.
Just iterate through dataset and create two instances of Sample for every row. Can be done with usage of COLLECTED_ATTRS, that 's fine.

Reply:
It is loop through range of number of samples, for each iteration there is loop through collected attributes ({attr_name: dataset generator, ...}). Sample() instance is created from next() value of datasets. So in fact there are more datasets and generator should provide next element of each one.

Reply:
I see. OK, concept is fine. Temporary creation of the dictionaries for every sample may be slow, but will see later.